File: eigenValue/codegen/mex/eigenValueSphere3/eigenValueSphere3.c
/*
 * Academic License - for use in teaching, academic research, and meeting
 * course requirements at degree granting institutions only.  Not for
 * government, commercial, or other organizational use.
 *
 * eigenValueSphere3.c
 *
 * Code generation for function 'eigenValueSphere3'
 *
 */

/* Include files */
#include "rt_nonfinite.h"
#include "eigenValueSphere3.h"
#include "eigenValueSphere3_emxutil.h"
#include "mpower.h"
#include "sum.h"
#include "triu.h"
#include "diag.h"
#include "blas.h"

/* Function Declarations */
static void euclidCoords(const real_T phi[399], real_T x[400]);

/* Function Definitions */
static void euclidCoords(const real_T phi[399], real_T x[400])
{
  real_T sinefact;
  int32_T i;
  x[0] = muDoubleScalarCos(phi[0]);
  sinefact = muDoubleScalarSin(phi[0]);
  for (i = 0; i < 398; i++) {
    x[i + 1] = muDoubleScalarCos(phi[i + 1]) * sinefact;
    sinefact *= muDoubleScalarSin(phi[i + 1]);
  }

  x[399] = sinefact;
}

void eigenValueSphere3(eigenValueSphere3StackData *SD, const real_T A[160000],
  real_T phi[399], real_T tol, real_T dt, emxArray_real_T *Vhist)
{
  real_T residual;
  real_T dA[400];
  real_T b_Vhist[2000];
  real_T u[400];
  real_T cpa;
  real_T Vd;
  char_T TRANSB;
  char_T TRANSA;
  real_T rowsums[400];
  ptrdiff_t m_t;
  ptrdiff_t n_t;
  ptrdiff_t k_t;
  ptrdiff_t lda_t;
  ptrdiff_t ldb_t;
  ptrdiff_t ldc_t;
  int32_T cnt;
  real_T sums_idx_0;
  real_T sums_idx_1;
  real_T sums_idx_2;
  real_T sums_idx_3;
  real_T sums_idx_4;
  int32_T i;
  real_T sum3fact;
  int32_T j;
  real_T ufact;
  real_T alpha;
  real_T dta;
  real_T cp;
  real_T sp;
  real_T cc;
  real_T ss;
  real_T sc;
  real_T V;
  real_T b_u[400];

  /*  Compute the lowest eigenvalue/vector of a symmetric matrix by minimizing the */
  /*  Rayleigh quotient over S^{n-1}. */
  residual = rtInf;
  diag(A, dA);
  memcpy(&SD->f0.AU[0], &A[0], 160000U * sizeof(real_T));
  triu(SD->f0.AU);
  memset(&b_Vhist[0], 0, 2000U * sizeof(real_T));
  euclidCoords(phi, u);
  cpa = 1.0;
  Vd = 0.0;
  TRANSB = 'N';
  TRANSA = 'N';
  memset(&rowsums[0], 0, 400U * sizeof(real_T));
  m_t = (ptrdiff_t)400;
  n_t = (ptrdiff_t)1;
  k_t = (ptrdiff_t)400;
  lda_t = (ptrdiff_t)400;
  ldb_t = (ptrdiff_t)400;
  ldc_t = (ptrdiff_t)400;
  dgemm(&TRANSA, &TRANSB, &m_t, &n_t, &k_t, &cpa, &A[0], &lda_t, &u[0], &ldb_t,
        &Vd, &rowsums[0], &ldc_t);
  for (cnt = 0; cnt < 400; cnt++) {
    rowsums[cnt] *= u[cnt];
  }

  b_Vhist[0] = sum(rowsums);

  /*  tic */
  sums_idx_0 = A[0] * mpower(u[0]);

  /*  i = k, j = k */
  sums_idx_1 = 0.0;

  /*  i < k, j = k */
  sums_idx_2 = 2.0 * (rowsums[0] - sums_idx_0);

  /*  i = k, j > k */
  sums_idx_3 = 0.0;

  /*  i < k, j > k */
  sums_idx_4 = (b_Vhist[0] - sums_idx_2) - sums_idx_0;

  /*  i > k, j > k */
  cpa = 1.0;
  Vd = 0.0;
  TRANSB = 'N';
  TRANSA = 'N';
  memset(&rowsums[0], 0, 400U * sizeof(real_T));
  m_t = (ptrdiff_t)400;
  n_t = (ptrdiff_t)1;
  k_t = (ptrdiff_t)400;
  lda_t = (ptrdiff_t)400;
  ldb_t = (ptrdiff_t)400;
  ldc_t = (ptrdiff_t)400;
  dgemm(&TRANSA, &TRANSB, &m_t, &n_t, &k_t, &cpa, &SD->f0.AU[0], &lda_t, &u[0],
        &ldb_t, &Vd, &rowsums[0], &ldc_t);
  for (cnt = 0; cnt < 400; cnt++) {
    rowsums[cnt] *= 2.0 * u[cnt];
  }

  i = 1;
  while ((i - 1 < 2000) && (!(residual < tol))) {
    sum3fact = 1.0;
    ufact = 1.0;
    for (j = 0; j < 399; j++) {
      alpha = 0.1;
      dta = dt / 0.1;
      cp = 1.0 / muDoubleScalarCos(phi[j]);
      sp = 1.0 / muDoubleScalarSin(phi[j]);
      cpa = muDoubleScalarCos(phi[j] + 0.1);
      residual = muDoubleScalarSin(phi[j] + 0.1);
      cc = cpa * cp;
      ss = residual * sp;
      sc = residual * cp;
      residual = cpa * sp;
      Vd = (((sums_idx_0 * (mpower(cc) - 1.0) + sums_idx_1 * (cc - 1.0)) +
             sums_idx_2 * (cc * ss - 1.0)) + sums_idx_3 * (ss - 1.0)) +
        sums_idx_4 * (mpower(ss) - 1.0);
      V = 0.1 + dta * Vd;
      cpa = rtInf;
      cnt = 0;
      while ((muDoubleScalarAbs(V) > 1.0E-13) && (cnt < 10) && (cpa - V > 0.0))
      {
        alpha -= V / ((1.0 + dta * ((((sums_idx_0 * (-2.0 * sc * cc) +
          sums_idx_1 * -sc) + sums_idx_2 * (cc * residual - sc * ss)) +
          sums_idx_3 * residual) + sums_idx_4 * 2.0 * residual * ss)) - dta * Vd
                      / alpha);
        dta = dt / alpha;
        cpa = muDoubleScalarCos(phi[j] + alpha);
        residual = muDoubleScalarSin(phi[j] + alpha);
        cc = cpa * cp;
        ss = residual * sp;
        sc = residual * cp;
        residual = cpa * sp;
        Vd = (((sums_idx_0 * (mpower(cc) - 1.0) + sums_idx_1 * (cc - 1.0)) +
               sums_idx_2 * (cc * ss - 1.0)) + sums_idx_3 * (ss - 1.0)) +
          sums_idx_4 * (mpower(ss) - 1.0);
        cpa = V;
        V = alpha + dta * Vd;
        cnt++;
      }

      u[j] = u[j] * cc * ufact;
      ufact *= ss;
      residual = u[j + 1] * ufact;
      sum3fact *= mpower(ss);
      sums_idx_0 = residual * A[(j + 400 * (j + 1)) + 1] * residual;

      /*  i = k, j = k */
      sums_idx_1 = 0.0;
      for (cnt = 0; cnt <= j; cnt++) {
        sums_idx_1 += A[cnt + 400 * (j + 1)] * u[cnt];
      }

      sums_idx_1 *= 2.0 * residual;
      sums_idx_3 = (cc * ss * sums_idx_2 + ss * sums_idx_3) - sums_idx_1;

      /*  i < k, j > k */
      sums_idx_2 = rowsums[j + 1] * sum3fact;
      sums_idx_4 = (ss * ss * sums_idx_4 - sums_idx_2) - sums_idx_0;

      /*  i > k, j > k */
      phi[j] += alpha;
    }

    u[399] *= ufact;
    cpa = 1.0;
    Vd = 0.0;
    TRANSB = 'N';
    TRANSA = 'N';
    memset(&rowsums[0], 0, 400U * sizeof(real_T));
    m_t = (ptrdiff_t)400;
    n_t = (ptrdiff_t)1;
    k_t = (ptrdiff_t)400;
    lda_t = (ptrdiff_t)400;
    ldb_t = (ptrdiff_t)400;
    ldc_t = (ptrdiff_t)400;
    dgemm(&TRANSA, &TRANSB, &m_t, &n_t, &k_t, &cpa, &SD->f0.AU[0], &lda_t, &u[0],
          &ldb_t, &Vd, &rowsums[0], &ldc_t);
    for (cnt = 0; cnt < 400; cnt++) {
      b_u[cnt] = u[cnt] * dA[cnt] * u[cnt];
      rowsums[cnt] *= 2.0 * u[cnt];
    }

    b_Vhist[i] = sum(rowsums) + sum(b_u);
    sums_idx_0 = A[0] * mpower(u[0]);

    /*  i = k, j = k */
    sums_idx_1 = 0.0;

    /*  i < k, j = k */
    sums_idx_2 = rowsums[0];

    /*  i = k, j > k */
    sums_idx_3 = 0.0;

    /*  i < k, j > k */
    sums_idx_4 = (b_Vhist[i] - rowsums[0]) - sums_idx_0;

    /*  i > k, j > k */
    residual = (b_Vhist[i - 1] - b_Vhist[i]) / muDoubleScalarAbs(b_Vhist[0]);
    i++;
  }

  /*  toc */
  j = 0;
  for (i = 0; i < 2000; i++) {
    if (b_Vhist[i] != 0.0) {
      j++;
    }
  }

  cnt = Vhist->size[0];
  Vhist->size[0] = j;
  emxEnsureCapacity((emxArray__common *)Vhist, cnt, sizeof(real_T));
  cnt = 0;
  for (i = 0; i < 2000; i++) {
    if (b_Vhist[i] != 0.0) {
      Vhist->data[cnt] = b_Vhist[i];
      cnt++;
    }
  }
}

/* End of code generation (eigenValueSphere3.c) */

File: eigenValue/eigenValueSphere3.m
% Compute the lowest eigenvalue/vector of a symmetric matrix by minimizing the
% Rayleigh quotient over S^{n-1}.

function Vhist = eigenValueSphere3(A,phi,tol,dt)
%#codegen
N = length(phi)+1;
Tmax = 2000;
residual = inf;
tolNewton =  1E-13;

dA = diag(A);
AU = triu(A,1);
Vhist = zeros(Tmax,1);
u = euclidCoords(phi);
rowsums = u.*(A*u);
Vhist(1) = sum(rowsums);
% tic
sums = zeros(1,5);
sums(1) = A(1,1)*u(1)^2;                          % i = k, j = k
sums(2) = 0;                                      % i < k, j = k
sums(3) = 2*(rowsums(1) - sums(1));               % i = k, j > k
sums(4) = 0;                                      % i < k, j > k
sums(5) = Vhist(1) - sums(3) - sums(1);           % i > k, j > k
trisum = 2*u.*(AU*u);

newtoniters = 0;
t = 0;
for i = 1:Tmax
    if  residual < tol
        break
    end
    sum3fact = 1;
    ufact = 1;
    for j = 1:N-1
    
        
        alpha = 0.1;
        dta = dt/alpha;
        cp = 1/cos(phi(j)); sp = 1/sin(phi(j));
        cpa = cos(phi(j)+alpha); spa = sin(phi(j)+alpha);
        cc = cpa*cp; ss = spa*sp; sc = spa*cp; cs = cpa*sp;
        
        Vd = sums(1)*(cc^2-1) + sums(2)*(cc-1) + sums(3)*(cc*ss - 1) + sums(4)*(ss-1) + sums(5)*(ss^2-1);
        V = alpha + dta*Vd;
        V_old = inf;
        cnt = 0;
        while abs(V) > tolNewton && cnt < 10 && V_old - V > 0
            dV = sums(1)*(-2*sc*cc) + sums(2)*(-sc) + sums(3)*(cc*cs - sc*ss) + sums(4)*cs + sums(5)*2*cs*ss;
            dV = 1+ dta*dV - dta*Vd/alpha;
            alpha = alpha - V/dV;
            dta = dt/alpha;
            cpa = cos(phi(j)+alpha); spa = sin(phi(j)+alpha);
            cc = cpa*cp; ss = spa*sp; sc = spa*cp; cs = cpa*sp;
            Vd = sums(1)*(cc^2-1) + sums(2)*(cc-1) + sums(3)*(cc*ss-1) + sums(4)*(ss-1) + sums(5)*(ss^2-1);
            V_old = V;
            V = alpha + dta*Vd;
            cnt = cnt + 1;
        end
        
        
        newtoniters = newtoniters + cnt;

        u(j) = u(j)*cc*ufact;
        ufact = ufact*ss;
        un = u(j+1)*ufact;
        sum3fact = sum3fact*ss^2;
        
        sums(1) = un*A(j+1,j+1)*un;              % i = k, j = k
        sums(2) = 0;

        for k = 1:j
            sums(2) = sums(2) + A(k,j+1)*u(k);
        end
        sums(2) = 2*un*sums(2);
        sums(4) = cc*ss*sums(3) + ss*sums(4) - sums(2);           % i < k, j > k
        sums(3) = trisum(j+1)*sum3fact;
        sums(5) = ss*ss*sums(5) - sums(3) - sums(1);           % i > k, j > k
        phi(j) = phi(j) + alpha;
    end
    u(end) = u(end)*ufact;
    trisum = 2*u.*(AU*u);

    dsum = sum(u.*dA.*u);
    Vhist(i+1) = sum(trisum)+dsum;
    sums(1) = A(1,1)*u(1)^2;                          % i = k, j = k
    sums(2) = 0;                                      % i < k, j = k
    sums(3) = trisum(1);                              % i = k, j > k
    sums(4) = 0;                                      % i < k, j > k
    sums(5) = Vhist(i+1) - sums(3) - sums(1);         % i > k, j > k
    residual = (Vhist(i) - Vhist(i+1))/abs(Vhist(1));
end
% toc
Vhist = Vhist(~~Vhist);
phi
end

function x = euclidCoords(phi)
n = length(phi);
x = zeros(n+1,1);
x(1) = cos(phi(1));
sinefact = sin(phi(1));
for i = 2:n
    x(i) = cos(phi(i))*sinefact;
    sinefact = sinefact*sin(phi(i));
end
x(end) = sinefact;
end